kvserver: add metric for replica circuit breaker

tbg opened the pull request:
Add a metric that tracks how many replicas have tripped circuit
breakers.
Add a metric that counts the trip events as well. This can highlight
problems where the circuit is tripped in error and immediately untrips
(which may not be caught by the first metric).

Also, as requested by @mwang1026, report the latter via telemetry as
well.

Fixes #74505.

Release note: None

tbg committed:
kvserver: add metric for replica circuit breaker

Add a metric that tracks how many replicas have tripped circuit
breakers.
Add a metric that counts the trip events as well. This can highlight
problems where the circuit is tripped in error and immediately untrips
(which may not be caught by the first metric).

Since tripped circuit breakers highlight an availability problem,
we're also adding an alert/aggregation rule.

Also, as requested by @mwang1026, report the count of trip events via
telemetry as well.

Fixes #74505.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74908)
<!-- Reviewable:end -->

tbg commented:
@aliher1911 would you mind reviewing this since @erikgrinaker is out? If there are any nontrivial questions feel free to call me up so we can chat about it.

erikgrinaker commented on pkg/kv/kvserver/metrics.go:
```diff
@@ -1459,6 +1479,10 @@ type StoreMetrics struct {
+
+	// Replica circuit breaker.
+	ReplicaCircuitBreakerCurTripped *metric.Gauge
+	ReplicaCircuitBreakerCumTripped *metric.Gauge
```
`CumTripped` should be a count, not a gauge.

erikgrinaker commented on pkg/kv/kvserver/metrics.go:
```diff
@@ -1242,6 +1242,26 @@ throttled they do count towards 'delay.total' and 'delay.enginebackpressure'.
+
+	// Replica circuit breaker.
+	metaReplicaCircuitBreakerCumTripped = metric.Metadata{
+		Name:        "kv.replica_circuit_breaker.num_trip_events",
```
nit: consider `num_tripped_events` for consistency.

erikgrinaker commented on pkg/kv/kvserver/replica_init.go:
```diff
@@ -128,7 +129,14 @@ func newUnloadedReplica(
 
-	r.breaker = newReplicaCircuitBreaker(store.cfg.Settings, store.stopper, r.AmbientContext, r)
+	r.breaker = newReplicaCircuitBreaker(
+		store.cfg.Settings, store.stopper, r.AmbientContext, r, func() {
+			telemetry.Inc(telemetryTripAsync)
+			r.store.Metrics().ReplicaCircuitBreakerCumTripped.Inc(1)
+			store.Metrics().ReplicaCircuitBreakerCurTripped.Inc(1)
+		}, func() {
+			store.Metrics().ReplicaCircuitBreakerCurTripped.Dec(1)
+		})
```
Consider either naming these closures, or adding a comment to clarify their purpose.

erikgrinaker reviewed (approved):
LGTM, modulo some minor comments.

tbg commented:
TFTR! Comments addressed, but looks like I have something to look into:

```
Error Trace:	client_replica_circuit_breaker_test.go:121
        	Error:      	Not equal: 
        	            	expected: int(0)
        	            	actual  : int64(-1)
        	Test:       	TestReplicaCircuitBreaker_LeaseholderTripped
```

I suspect we're not always emitting the correct number of "Reset" events.

tbg committed:
circuit: don't launch probe when breaker not tripped

It was possible for probes to be launched when the breaker was not
tripped. This generated `OnReset` events that were not accompanied
by preceding `OnTrip(nil, err)` events. We're trying to use these
events to keep track of how many breakers are tripped, so it makes
sense to have them pair up.

With the new invariant checks added, this failed immediately under
stress prior to this commit, in `TestBreakerRealistic`.

Release note: None

tbg commented:
Got the sucker, prepended a commit.

erikgrinaker reviewed (approved):
Nice, thanks.

tbg commented:
bors r=erikgrinaker
